lib/widgets/compose_box.dart

```diff
@@ -269,19 +269,101 @@ class ComposeContentController extends ComposeController<ContentValidationError>
   }
 }
 
-class _ContentInput extends StatelessWidget {
+class _ContentInput extends StatefulWidget {
   const _ContentInput({
     required this.narrow,
+    required this.destination,
     required this.controller,
     required this.focusNode,
     required this.hintText,
   });
 
   final Narrow narrow;
+  final SendableNarrow destination;
   final ComposeContentController controller;
   final FocusNode focusNode;
   final String hintText;
 
+  @override
+  State<_ContentInput> createState() => _ContentInputState();
+}
+
+class _ContentInputState extends State<_ContentInput> with WidgetsBindingObserver {
+  @override
+  void initState() {
+    super.initState();
+    widget.controller.addListener(_contentChanged);
+    widget.focusNode.addListener(_focusChanged);
+    WidgetsBinding.instance.addObserver(this);
+  }
+
+  @override
+  void didUpdateWidget(covariant _ContentInput oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    if (widget.controller != oldWidget.controller) {
+      oldWidget.controller.removeListener(_contentChanged);
+      widget.controller.addListener(_contentChanged);
+    }
+    if (widget.focusNode != oldWidget.focusNode) {
+      oldWidget.focusNode.removeListener(_focusChanged);
+      widget.focusNode.addListener(_focusChanged);
+    }
+  }
+
+  @override
+  void dispose() {
+    widget.controller.removeListener(_contentChanged);
+    widget.focusNode.removeListener(_focusChanged);
+    WidgetsBinding.instance.removeObserver(this);
+    super.dispose();
+  }
+
+  void _contentChanged() {
+    final store = PerAccountStoreWidget.of(context);
+    (widget.controller.text.isEmpty)
+      ? store.typingNotifier.stoppedComposing()
+      : store.typingNotifier.keystroke(widget.destination);
+  }
+
+  void _focusChanged() {
+    if (widget.focusNode.hasFocus) {
+      // Content input getting focus doesn't necessarily mean that
+      // the user started typing, so do nothing.
+      return;
+    }
+    // Losing focus usually indicates that the user has navigated away or
+    // clicked on other UI elements.  On Android, this does not get triggered
+    // unless the user navigates away from the page or focuses on another input.
+    // See: https://github.com/zulip/zulip-flutter/pull/897#discussion_r1818188643
+    final store = PerAccountStoreWidget.of(context);
+    store.typingNotifier.stoppedComposing();
```

Comment:
```suggestion
    final store = PerAccountStoreWidget.of(context);
    store.typingNotifier.stoppedComposing();
```

The "On Android" part doesn't seem relevant to understanding this code; it's not part of why we wrote this code the way we did. (If hiding the keyboard did unfocus the input, we'd be glad and would still want to say the user has stopped typing.) So it can be left out.

I'm also not sure that comment is accurate — it tries to give a complete list of the things that will cause unfocus, and I'm not sure there aren't other things we haven't thought of that also do so. (What about opening the bottom sheet; does that unfocus the input? I suspect it does.)

Then the "Losing focus" sentence seems like it's just explaining the meaning of focus.